### PR TITLE
Plone 5 support + blacklist fix

### PR DESCRIPTION
**Pull request by thet**

### CODE REVIEW

**Comment by hvelarde** [comment]
@thet can't we use this new feature instead? https://github.com/plone/Products.CMFPlone/issues/1406

File: src/collective/lazysizes/transform.py [hunk 1]
```diff
@@ -16,7 +16,7 @@
 ROOT_SELECTOR = '//*[@id="content"]'
 
 # elements by CSS class; http://stackoverflow.com/a/1604480
-CLASS_SELECTOR = '//*[contains(concat(" ", normalize-space(@class), " "), " {0} ")]'
+CLASS_SELECTOR = '//*[contains(@class, "{0}")]'
```
Comment (hvelarde) [inline src/collective/lazysizes/transform.py]: this change needs tests.

File: src/collective/lazysizes/profiles/plone5/registry.xml [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    <value key="merge_with">default</value>
+    <value key="depends"></value>
+    <value key="compile">False</value>
+    <value key="enabled">True</value>
+    <value key="resources">
+      <element>lazysizes</element>
+    </value>
+  </records>
+
+  <!-- NOTE: if you use ``lazysizes-twitter``, disable the ``lazysizes`` bundle -->
```
Comment (hvelarde) [inline src/collective/lazysizes/profiles/plone5/registry.xml]: what do you mean with this?

File: src/collective/lazysizes/profiles/base/metadata.xml [hunk 3]
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0"?>
 <metadata>
-  <version>7</version>
```
Comment (hvelarde) [inline src/collective/lazysizes/profiles/base/metadata.xml]: why do we need to change the profile?

File: CHANGES.rst [hunk 4]
```diff
@@ -1,8 +1,15 @@
...
-3.0.1 (unreleased)
-------------------
+3.1 (unreleased)
+----------------
+
+- Plone 5 + Plone 4 compatibility via two separate profiles.
+  The correct Plone 4 / Plone 5 is registered under the name ``default`` and automatically applied on installation.
+  [thet]
+
+- Fix the xpath class selector so that blackklisted classes are really excluded from the transform.
```
Comment (hvelarde) [inline CHANGES.rst]: "blacklisted"

File: src/collective/lazysizes/configure.zcml [hunk 5]
```diff
@@ -21,11 +21,18 @@
       permission="collective.lazysizes.Setup"
       />
 
+  <!-- zope resource files -->
   <browser:resourceDirectory
       name="collective.lazysizes"
       directory="static"
-      layer="collective.lazysizes.interfaces.ILazySizesLayer"
```
Comment (hvelarde) [inline src/collective/lazysizes/configure.zcml]: is this because a bug in Plone 5?

File: src/collective/lazysizes/profiles.zcml [hunk 6]
```diff
@@ -2,15 +2,36 @@
     xmlns="http://namespaces.zope.org/zope"
     xmlns:genericsetup="http://namespaces.zope.org/genericsetup"
     xmlns:i18n="http://namespaces.zope.org/i18n"
+    xmlns:zcml="http://namespaces.zope.org/zcml"
     i18n_domain="collective.lazysizes">
 
   <genericsetup:registerProfile
+      name="base"
```
Comment (hvelarde) [inline src/collective/lazysizes/profiles.zcml]: can't we use this new feature instead? https://github.com/plone/Products.CMFPlone/issues/1406

File: src/collective/lazysizes/transform.py (hunk 1 above)
Comment (thet) [reply src/collective/lazysizes/transform.py]: I'm currently working on it

File: src/collective/lazysizes/profiles/base/metadata.xml (hunk 3 above)
Comment (thet) [reply src/collective/lazysizes/profiles/base/metadata.xml]: because the profile has changed and should be re-loaded.

on the other hand - for plone 4 nothing is changed...

File: src/collective/lazysizes/configure.zcml (hunk 5 above)
Comment (thet) [reply src/collective/lazysizes/configure.zcml]: first i tried to use the ++resource++... resources, which does not work with bundles (they need to be able to be stored in the ZODB) - therefore I switched to plone.resource directories. so this change might not be needed, but doesn't hurt also. e.g. the controlpanel references a image via the resource directory path...

File: src/collective/lazysizes/profiles.zcml (hunk 6 above)
Comment (thet) [reply src/collective/lazysizes/profiles.zcml]: Ha, cool - I actually didn't know this.
But still - I'd prefer to not import the jsregistry.xml.
On the other hand, this might not hurt that much...

File: src/collective/lazysizes/profiles/plone5/registry.xml (hunk 2 above)
Comment (thet) [reply src/collective/lazysizes/profiles/plone5/registry.xml]: i made two bundles - lazysizes and lazysizes-twitter. the twitter one also loads lazysizes AFTER the twitter code. to not load two lazysizes, you have to disable one of them.

not so elegant, but maybe the simplest solution.

the twitter code should be loaded before lazysizes itself - so say the docs. with bundles you can only define dependencies, but no insert-before. we cannot let lazysizes depend on lazysizes-twitter, because the latter might just not be loaded (as it's the default). therefore how it's done here.

**Comment by thet** [comment]
@hvelarde regardin updates: I do not see a problem with that. You just replace the files. If the paths have changed, you change it ins registry.xml too. if ``last_compilation`` is empty, the non-compileable bundle will be cooked after restart or when the registry is set in dev mode and back. otherwise you just update ``last_compilation`` with a new ISO date string.

**Comment by thet** [comment]
I'll work on the issues and update this PR soon.

File: src/collective/lazysizes/configure.zcml (hunk 5 above)
Comment (hvelarde) [reply src/collective/lazysizes/configure.zcml]: this is a know bug in Plone 5: https://github.com/plone/Products.CMFPlone/issues/1749

File: src/collective/lazysizes/transform.py (hunk 1 above)
Comment (rodfersou) [reply src/collective/lazysizes/transform.py]: you are not considering when there are more than one class in the same element..

please leave it as it was before.

Comment (thet) [reply src/collective/lazysizes/transform.py]: ah - i understand the issue. good to know for the exact match trick, i always used just ``contains``.

i had a problem with not filtering the blacklisted classes, so i provided this "fix". now i have to re-investigate that problem.

Comment (thet) [reply src/collective/lazysizes/transform.py]: But it should be filtered with the root selector. It should read:

```CLASS_SELECTOR = ROOT_SELECTOR + '//*[contains(concat(" ", normalize-space(@class), " "), " {0} ")]```

Comment (thet) [reply src/collective/lazysizes/transform.py]: Of course, you're already doing it.

**Comment by thet** [comment]
@rodfersou this commit solves the issue in my case, where ``img`` elements with the blacklist class directly set on it wasn't found:
https://github.com/collective/collective.lazysizes/pull/36/commits/4c2faf0f6a82c545977db7756a5b58eedd60261b

File: src/collective/lazysizes/profiles/plone5/registry.xml (hunk 2 above)
Comment (hvelarde) [reply src/collective/lazysizes/profiles/plone5/registry.xml]: but why do we want 2 different bundles? how do you select or the other? you need to explain better this not only for us, but for the rest of the integrators; a note on the README should be added in case we decide that 2 bundles are needed indeed.

File: src/collective/lazysizes/profiles.zcml (hunk 6 above)
Comment (hvelarde) [reply src/collective/lazysizes/profiles.zcml]: I prefer to move the registration of JS in Plone 4 to a setup handler in Python code then because this is causing other issues as you can see on tests.

File: src/collective/lazysizes/profiles/plone5/registry.xml [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    xmlns:i18n="http://xml.zope.org/namespaces/i18n">
+
+  <!--
+    BUNDLES - ready to use
+    The compiled bundle files are empty - they are generated/cooked by Plone.
+  -->
+
+  <records prefix="plone.bundles/lazysizes" interface="Products.CMFPlone.interfaces.IBundleRegistry">
+    <value key="last_compilation"></value>
+    <value key="jscompilation">++plone++collective.lazysizes/bundle-lazysizes-compiled.min.js</value>
```
Comment (hvelarde) [inline src/collective/lazysizes/profiles/plone5/registry.xml]: I don't understand how this works, you're adding empty files in both cases and that makes no sense to me.

File: src/collective/lazysizes/transform.py [hunk 8]
```diff
@@ -141,7 +141,7 @@ def _blacklist(self, result, blacklisted_classes):
 
         path = []
         for css_class in blacklisted_classes:
-            path.append('{0}{1}//img|{0}{1}//iframe|{0}{1}//blockquote'.format(
+            path.append('{0}{1}|{0}{1}//img|{0}{1}//iframe|{0}{1}//blockquote'.format(
```
Comment (hvelarde) [inline src/collective/lazysizes/transform.py]: this change needs a test.

**Comment by hvelarde** [comment]
@thet sorry for making so many questions but this need to be completely clear for me or it will break as soon as I'll make a new release.

File: src/collective/lazysizes/profiles/plone5/registry.xml (hunk 7 above)
Comment (thet) [reply src/collective/lazysizes/profiles/plone5/registry.xml]: I understand this doesn't make too much sense. But this is the reason:
The resource registry expects these bundle files to be existent, but immediately cooks a new bundle file from it's referenced resources if the ``last_compilation`` time isn't set and ``compile`` is set to ``False``.
Better would be to fix the resource registry to not expect a bundle compilation file here and just create one by itself in that case.
The file is empty because it's cooked (concat + minified) anyways. If we would provide a file by ourself, we had to maintain it manually and upgrade it with every upgrade to the lazysizes core file. The cooked bundle is almost the same like the lazysizes core file, except that a requirejs undefine statement is added to the top of the file and a requirejs redefine is added to the end - this is to isolate legacy js from the global requirejs definitions (otherwise we would get a "Mismatched anonymous define" error).
Does it make more sense now?

File: src/collective/lazysizes/transform.py (hunk 8 above)
Comment (thet) [reply src/collective/lazysizes/transform.py]: I can provide a test.
The interface description for ``css_class_blacklist`` says:
""" [...] elements with that class directly applied to them, or to a parent element, will be skipped."""
https://github.com/collective/collective.lazysizes/blob/master/src/collective/lazysizes/interfaces.py#L23

I actually also expected the blacklist class being able to set on the img elements directly, even before reading the interface description :)

**Comment by thet** [comment]
@hvelarde no problem for the questions, that's what the review is for. Thanks for the review!
By the way - I like the approach using plone.transformchain to add this lazysizes feature!

File: src/collective/lazysizes/configure.zcml (hunk 5 above)
Comment (thet) [reply src/collective/lazysizes/configure.zcml]: Commented on that: https://github.com/plone/Products.CMFPlone/issues/1749#issuecomment-332007871

File: src/collective/lazysizes/profiles/plone5/registry.xml (hunk 7 above)
Comment (hvelarde) [reply src/collective/lazysizes/profiles/plone5/registry.xml]: there are a couple of thing we wanted to do and maybe it's time to do it before your fix:

* use Webpack to build the resources of this package (CC @rodfersou)
* update to new lazysizes version which seems to use CommonJS format for plugins

regarding the later, I found this on the RequireJS documentation: http://www.requirejs.org/docs/commonjs.html

we can check if it's better to use the UMD versions of lazysizes or not.

**Comment by thet** [comment]
Some changes:
- Re-combined separate plone5 and plone4 profiles to only use a ``default`` profile. Note: I cannot reproduce the error mentioned in https://github.com/plone/Products.CMFPlone/issues/1749 event with a fresh install. Someone else?
- Extended the ``uninstall`` profile with more de-registrations.
- No separate lazysizes and lazysizes-twitter bundles any more. The twitter JavaScript adds only a very small overhead, so it's delivered by default. Who ever does not want this can do a custom bundle registration.
- metadata version restored to ``7``
- Plone 5 resources are imported conditionally.

Changes in the transformer:
- Fix to exclude blacklist classes directly set on the element itself.
- Added ``available`` method, so that subclasses of the transformer can do their own available logic. E.g. by allowing also non-anonymous users have the transform.

**Comment by thet** [comment]
@hvelarde not sure, if it's needed to wait for webpack. in the end, only a compiled resource will be produced, which has to be included as i did (AFAIK).

File: CHANGES.rst [hunk 9]
```diff
@@ -1,8 +1,20 @@
...
+3.1 (unreleased)
+----------------
+
+- Plone 5 + Plone 4 compatibility.
+  [thet]
+
+- Extend the ``uninstall`` profile with more de-registrations.
+  [thet]
+
+- Add a ``available`` method to the transform, which can be overridden to e.g. allow the transform also for non-anonymous users.
```
Comment (hvelarde) [inline CHANGES.rst]: I would prefer you to help me reviewing #35; I prepared it for @gforcada, but it seems he was not very excited about it ;-)

File: src/collective/lazysizes/profiles/uninstall/controlpanel.xml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+<object name="portal_controlpanel">
+  <configlet
+      remove="True"
+      title="lazysizes"
+      action_id="lazysizes"
+      appId="collective.lazysizes"
```
Comment (hvelarde) [inline src/collective/lazysizes/profiles/uninstall/controlpanel.xml]: I think you only need the `action_id` here.

File: src/collective/lazysizes/profiles/uninstall/registry.xml [hunk 11]
```diff
@@ -1,4 +1,13 @@
...
 <registry>
+  <!-- settings -->
   <records interface="collective.lazysizes.interfaces.ILazySizesSettings" remove="true" />
+  <!-- Plone 5 - resources -->
+  <records condition="have plone-5" remove="True" prefix="plone.bundles/lazysizes" interface="Products.CMFPlone.interfaces.IBundleRegistry"/>
+  <records condition="have plone-5" remove="True" prefix="plone.resources/lazysizes" interface='Products.CMFPlone.interfaces.IResourceRegistry'/>
+  <records condition="have plone-5" remove="True" prefix="plone.resources/lazysizes-twitter" interface='Products.CMFPlone.interfaces.IResourceRegistry'/>
+  <!-- Plone 5 - plone bundle does not depend on any other per default -->
+  <records condition="have plone-5" prefix="plone.bundles/plone" interface="Products.CMFPlone.interfaces.IBundleRegistry" purge="False">
+    <value key="depends"></value>
```
Comment (hvelarde) [inline src/collective/lazysizes/profiles/uninstall/registry.xml]: are you sure this is right? what happens if there are other add-ons installed?

File: src/collective/lazysizes/transform.py [hunk 12]
```diff
@@ -9,14 +9,14 @@
 
 
 # grey rectangule, 16x16
-PLACEHOLDER = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAA1BMVEXy8vJkA4prAAAAC0lEQVQI12MgEQAAADAAAWV61nwAAAAASUVORK5CYII='
+PLACEHOLDER = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQAQMAAAAlPW0iAAAAA1BMVEXy8vJkA4prAAAAC0lEQVQI12MgEQAAADAAAWV61nwAAAAASUVORK5CYII='  # noqa
```
Comment (hvelarde) [inline src/collective/lazysizes/transform.py]: I prefer to use the more specific `# noqa: E501` here

File: src/collective/lazysizes/transform.py [hunk 13]
```diff
@@ -9,14 +9,14 @@
...
 
 # to avoid additional network round trips to render content above the fold
 # we only process elements inside the "content" element
 ROOT_SELECTOR = '//*[@id="content"]'
 
 # elements by CSS class; http://stackoverflow.com/a/1604480
-CLASS_SELECTOR = '//*[contains(concat(" ", normalize-space(@class), " "), " {0} ")]'
+CLASS_SELECTOR = '//*[contains(concat(" ", normalize-space(@class), " "), " {0} ")]'  # noqa
```
Comment (hvelarde) [inline src/collective/lazysizes/transform.py]: I prefer to use the more specific `# noqa: E501` here

File: src/collective/lazysizes/transform.py [hunk 14]
```diff
@@ -141,20 +141,24 @@ def _blacklist(self, result, blacklisted_classes):
 
         path = []
         for css_class in blacklisted_classes:
-            path.append('{0}{1}//img|{0}{1}//iframe|{0}{1}//blockquote'.format(
+            path.append('{0}{1}|{0}{1}//img|{0}{1}//iframe|{0}{1}//blockquote'.format(  # noqa
```
Comment (hvelarde) [inline src/collective/lazysizes/transform.py]: I prefer to use the more specific `# noqa: E501` here

File: src/collective/lazysizes/tests/test_transformer.py [hunk 15]
```diff
@@ -88,3 +95,91 @@ def test_lazyload_tweet_no_script(self):
...
+            + u'</div></body></html>'
+        url = u'http://example.com/foo.png'
+
+        logout()  # Make available
+
+        # Case 1: Do not transform Blacklisted - single class
+        html = tpl.format(url, 'nolazyload')
+        result = self.transformer.transformIterable(html, 'utf-8')
+        img = result.tree.xpath('//img')[0]
+        self.assertTrue('data-src' not in img.attrib.keys())
```
Comment (hvelarde) [inline src/collective/lazysizes/tests/test_transformer.py]: use `self.assertNotIn()` instead.

File: src/collective/lazysizes/tests/test_transformer.py [hunk 16]
```diff
@@ -88,3 +95,91 @@ def test_lazyload_tweet_no_script(self):
...
+        self.assertEqual(img.attrib['src'], url)
+
+        # Case 2: Do not transform Blacklisted - multiple classes
+        html = tpl.format(url, 'nolazyload secondclass thirdclass')
+        result = self.transformer.transformIterable(html, 'utf-8')
+        img = result.tree.xpath('//img')[0]
+        self.assertTrue('data-src' not in img.attrib.keys())
```
Comment (hvelarde) [inline src/collective/lazysizes/tests/test_transformer.py]: use `self.assertNotIn()` instead.

File: src/collective/lazysizes/profiles/uninstall/controlpanel.xml (hunk 10 above)
Comment (thet) [reply src/collective/lazysizes/profiles/uninstall/controlpanel.xml]: I thought so too, but under Plone 4 you need all of these. (100% sure for the title, not 100% for the appId).
Plone 5 works only with action_id

File: src/collective/lazysizes/profiles/uninstall/registry.xml (hunk 11 above)
Comment (thet) [reply src/collective/lazysizes/profiles/uninstall/registry.xml]: Yes, that might be a problem.
We can also leave out changing the plone bundle's dependency while installing in the first place - and leave it up to the integrator to do this little optimization. I did not do any tests but I think it's better to have lazysizes loaded as first javascript, so it can do it's works as soon as possible.

File: src/collective/lazysizes/tests/test_transformer.py (hunk 15 above)
Comment (thet) [reply src/collective/lazysizes/tests/test_transformer.py]: I actually prefer assertTrue because the evaluation expression is easier to read.
But I understand, the test failure output is less informative.
Will change with next iteration.

File: src/collective/lazysizes/transform.py (hunk 12 above)
Comment (thet) [reply src/collective/lazysizes/transform.py]: you're right, but this goes against my developer laziness. I'll change it but I refuse to remember the E501 error code ;)

File: CHANGES.rst (hunk 9 above)
Comment (thet) [reply CHANGES.rst]: ok, there is another PR - didn't check for it.
an option makes totally sense - maybe in combination with the ``available`` method done by me. i could think of other use cases to disable the transform - e.g. on a per content type basis.

File: src/collective/lazysizes/transform.py (hunk 12 above)
Comment (hvelarde) [reply src/collective/lazysizes/transform.py]: the problem with using just `# noqa` is that it will hide other issues like, for instance, if we decide to later not to use this constant.

File: src/collective/lazysizes/tests/test_transformer.py (hunk 15 above)
Comment (hvelarde) [reply src/collective/lazysizes/tests/test_transformer.py]: in fact is the opposite: using `assertTrue` forces you to first evaluate the assertion and then evaluate the expression itself; that's a two-step mental process; see:

```python
self.assertTrue('data-src' not in img.attrib.keys())
self.assertNotIn('data-src', img.attrib)  # I think there's no need for .keys() neither
```

same happens with this, for instance:

```python
self.assertTrue(foo is None)
self.assertIsNone(foo)
```

File: CHANGES.rst (hunk 9 above)
Comment (thet) [reply CHANGES.rst]: Reviewed!

Comment (thet) [reply CHANGES.rst]: Actually - I guess I'll remove the commit with factoring out the ``available`` method.

**Comment by thet** [comment]
Ready for re-review!

File: CHANGES.rst [hunk 17]
```diff
@@ -1,8 +1,20 @@
...
+  [thet]
+
+- Fix the blacklist xpath selector to also match elements with the class directly set on it.
```
Comment (hvelarde) [inline CHANGES.rst]: I think it would be sane to split this into a different PR so it will be easier for you to do the rebase now; we can also get rid of the functional tests there.

File: README.rst [hunk 18]
```diff
@@ -57,6 +57,8 @@ Go to the 'Site Setup' page in a Plone site and click on the 'Add-ons' link.
 
 Check the box next to ``collective.lazysizes`` and click the 'Activate' button.
 
+You should make sure, that the lazysizes JavaScript is loaded early in order to be able to do it's work as soon as possible. It would not work well if you have the resource loaded at the end of the HTML document.
```
Comment (hvelarde) [inline README.rst]: I don't understand this statement; do you mind to elaborate?

File: src/collective/lazysizes/tests/test_transformer.py [hunk 19]
```diff
@@ -88,3 +95,91 @@ def test_lazyload_tweet_no_script(self):
         element = lxml.html.fromstring(TWEET_NO_SCRIPT)
         # the transformer returns None (skip element)
         self.assertIsNone(self.transformer._lazyload_tweet(element))
+
+    def test_lazyload_availability(self):
```
Comment (hvelarde) [inline src/collective/lazysizes/tests/test_transformer.py]: this is already on master now and I split the test to make it simpler:

https://github.com/collective/collective.lazysizes/blob/master/src/collective/lazysizes/tests/test_transformer.py#L54-L65

File: src/collective/lazysizes/tests/test_transformer.py [hunk 20]
```diff
@@ -88,3 +95,91 @@ def test_lazyload_tweet_no_script(self):
...
+        # NOTE: You should subclass and register the transform, not monkey
+        # patching like here.
+        transformer.available = lambda: True
+        result = transformer.transformIterable(html, 'utf-8')
+        img = result.tree.xpath('//img')[0]
+        self.assertEqual(img.attrib['data-src'], url)
+        self.assertEqual(img.attrib['src'], PLACEHOLDER)
+
+    def test_lazyload_blacklist(self):
+        """Test, if blacklisted elements are transformed.
```
Comment (hvelarde) [inline src/collective/lazysizes/tests/test_transformer.py]: we must get rid of this tests if all cases are now covered here:

https://github.com/collective/collective.lazysizes/blob/master/src/collective/lazysizes/tests/test_lazysizes.py

File: README.rst (hunk 18 above)
Comment (thet) [reply README.rst]: What I mean is, that if the lazyload JS is loaded at the very end of the HTML document, then all images would already been loaded and lazyload could not do it's work.
as long as it's loaded in the HTML head section, all should be good, though. Although I configured the lazyload JS in the project where I use lazyload to be loaded BEFORE the ``plone`` bundle - probably that doesn't have much effect.

Do you have a suggestion for a clearer statement or should I just drop this paragraph?

**Comment by thet** [comment]
@hvelarde rebased and ready to review

File: README.rst (hunk 18 above)
Comment (hvelarde) [reply README.rst]: in fact that is not true as the transform is already in place no matter where you place the lazysizes code; just as a test, visit any of these sites with JS disabled:

* http://www.brasil2016.gov.br/pt-br
* https://www.cartacapital.com.br/
* https://www.conversaafiada.com.br/

as you can see no image is loaded at all besides the site logo.

I suggest just to remove this paragraph.

Comment (thet) [reply README.rst]: You're right. It's the transform which makes the src attr into a data-src something. I should have known. Will remove that paragraph.
... Still it makes sense to load it before any other JS, maybe...?

**Comment by thet** [comment]
\+ removed the misleading README.rst paragraph
